src/librustc/hir/map/def_collector.rs

```diff
@@ -268,34 +269,41 @@ impl<'a> visit::Visitor<'a> for DefCollector<'a> {
     fn visit_expr(&mut self, expr: &'a Expr) {
         let parent_def = self.parent_def;
 
-        match expr.node {
+        self.parent_def = match expr.node {
```

Comment:
can you store this in a temporary variable and assign the `self.parent_def = temp.or(self.parent_def)` below? I think it would make the steps involved here a bit clearer